### ISSUE TITLE
issue(tun/lwip): page fault when closing a tun device

### ISSUE THREAD

**Issue by rinor** [nanovms/nanos#2061]
I'll prepare and post some code to reproduce this in a few, but this is the page fault that's giving me grief atm
```c
cpu: 0, context type: 2
lastvector: 000000000000000e (Page fault)
     frame: ffffc00002820000
      type: syscall
active_cpu: 0000000000000000
 stack top: ffffc00002827ff0
error code: 0000000000000000
   address: 0000000000000090

   rax: 0000000000000001
   rbx: ffffffffffffffe8
   rcx: ffffc00000a04b58
   rdx: 0000000000000003
   rsi: ffffc0000020cc00
   rdi: 0000000000000001
   rbp: ffffc00002827db8
   rsp: ffffc00002827d98
    r8: ffffc00001c00200
    r9: ffffc0000020cc00
   r10: 0000000000000b00
   r11: 0000000000000001
   r12: ffffc00000a04a00
   r13: ffffc00000a04a00
   r14: ffffc00002400004
   r15: 00000000000000ff
   rip: ffffffff83b6e610	(tun_if_output + 0000000000000060/00000000000001dd)
rflags: 0000000000010097
    ss: 0000000000000008
    cs: 0000000000000010
    ds: 0000000000000000
    es: 0000000000000000
fsbase: 0000000000000000
gsbase: 0000000000000000

frame trace:
ffffc00002827dc0:   ffffffffcacd431e	(ip4_output_if + 000000000000018e/000000000000021f)
ffffc00002827e20:   ffffffffcace7255	(tcp_output_control_segment + 00000000000000b5/000000000000017a)
ffffc00002827e80:   ffffffffcaceaaf0	(tcp_abandon + 00000000000002f0/0000000000000328)
ffffc00002827ef0:   ffffffffcaceabef	(tcp_netif_ip_addr_changed_pcblist + 000000000000008f/0000000000000144)
ffffc00002827f20:   ffffffffcaceba08	(tcp_netif_ip_addr_changed + 0000000000000058/00000000000001bf)
ffffc00002827f40:   ffffffffcace2b46	(netif_remove + 0000000000000246/000000000000028a)
ffffc00002827f90:   ffffffff83b6e894	(tun_close + 0000000000000104/0000000000000131)
ffffc00002827fc0:   ffffffffcac7e7b8	(syscall_handler + 0000000000000278/00000000000004d0)

kernel load offset ffffffffca9ad000

loaded klibs: tun@0xffffffff83b6e000/0x4000 

stack trace:
ffffc00002827d98:   0000000000000006
ffffc00002827da0:   ffffc00001c00200
ffffc00002827da8:   ffffc0000240001c
ffffc00002827db0:   ffffc00000a04a00
ffffc00002827db8:   ffffc00002827e18
ffffc00002827dc0:   ffffffffcacd431e	(ip4_output_if + 000000000000018e/000000000000021f)
ffffc00002827dc8:   ffffc00002827dd8
ffffc00002827dd0:   0000000000000006
ffffc00002827dd8:   ffffc00001c00200
ffffc00002827de0:   61c457e82695cec4
ffffc00002827de8:   ffffc00002827e28
ffffc00002827df0:   ffffc0000240001c
ffffc00002827df8:   ffffc00001c00200
ffffc00002827e00:   ffffc00002400004
ffffc00002827e08:   ffffc00000a04a00
ffffc00002827e10:   ffffc00002400000
ffffc00002827e18:   ffffc00002827e78
ffffc00002827e20:   ffffffffcace7255	(tcp_output_control_segment + 00000000000000b5/000000000000017a)
ffffc00002827e28:   ffffc00000a04a00
ffffc00002827e30:   ffffffffcace7214	(tcp_output_control_segment + 0000000000000074/000000000000017a)
ffffc00002827e38:   ffffc00002400004
ffffc00002827e40:   ffffc00001c00256
ffffc00002827e48:   0000000000000001
ffffc00002827e50:   ffffc00002400000
ffffc00002827e58:   0000000000000001
ffffc00002827e60:   ffffffffcac10a90	(lwip_tcp_conn_err + 0000000000000000/0000000000000147)
ffffc00002827e68:   000000000000e79d
ffffc00002827e70:   000000000000000e
ffffc00002827e78:   ffffc00002827ee8
ffffc00002827e80:   ffffffffcaceaaf0	(tcp_abandon + 00000000000002f0/0000000000000328)
ffffc00002827e88:   0000000000000050
ffffc00002827e90:   ffffffffcacea982	(tcp_abandon + 0000000000000182/0000000000000328)
```

**Comment by rinor** [comment]
Pushed the code at https://github.com/rinor/wireguard-netstack-nanos

- **Server** crash

```sh
# run server
ops run -c config/klib_server_http.json bin/klib_server_http --smp=1

# run client 3-4 times in sequential
ops run -c config/klib_client_http.json bin/klib_client_http --smp=1
# wait for it to complete

ops run -c config/klib_client_http.json bin/klib_client_http --smp=1
# wait for it to complete

ops run -c config/klib_client_http.json bin/klib_client_http --smp=1
# wait for it to complete

# stop server
bash scripts/qemu_powerdown_server.sh
```

It might take a few attempts, but eventually it will crash because of this line https://github.com/rinor/wireguard-netstack-nanos/blob/ddb7db62a12356c6aff75d88af9b587a31297dc1/cmd/nanos_klib/server_http/server_http.go#L123. 

During those runs the client may or may not crash.

- **Client** crash

Change config at https://github.com/rinor/wireguard-netstack-nanos/blob/ddb7db62a12356c6aff75d88af9b587a31297dc1/config/klib_client_http.json#L9

from 

```json
"CLEANUP_ACTIONS_ORDERED": "HTTP_CLIENT_CLOSE_IDLE_CONNECTIONS WG_DEVICE_DOWN WG_DEVICE_CLOSE",
```

to one of the following:

```json
"CLEANUP_ACTIONS_ORDERED": "WG_DEVICE_DOWN WG_DEVICE_CLOSE",
```

```json
"CLEANUP_ACTIONS_ORDERED": "HTTP_CLIENT_CLOSE_IDLE_CONNECTIONS WG_DEVICE_CLOSE",
```

```json
"CLEANUP_ACTIONS_ORDERED": "WG_DEVICE_CLOSE",
```

and eventually it will blow because of this line https://github.com/rinor/wireguard-netstack-nanos/blob/ddb7db62a12356c6aff75d88af9b587a31297dc1/cmd/nanos_klib/client_http/client_http.go#L115

the following config should disable that call for testing

```json
"CLEANUP_ACTIONS_ORDERED": "HTTP_CLIENT_CLOSE_IDLE_CONNECTIONS WG_DEVICE_DOWN",
```

**Comment by rinor** [comment]
- `ops profile`

```sh
Ops version: 0.1.42-11-g69f3669
Nanos version: 0.1.52-17-g7430b4d (feature/map_growsdown)
Qemu version: 7.2.13
OS: linux
Arch: amd64
Virtualized: false

```

**Comment by francescolavra** [comment]
@rinor thanks for providing all the details to reproduce this issue. Fixed in #2063.

**Comment by rinor** [comment]
tested and works as expected, thank you.
